pkg/controller/controller.go

```diff
@@ -177,96 +194,163 @@ func (tc *TenantController) processItem(e Event) error {
 	// process events based on its type
 	switch e.eventType {
 	case Create:
-		return tc.syncResources(e.tenantId)
+		return tc.onTenantCreate(e.tenantId)
+	case Update:
+		return tc.onTenantUpdate(e.tenantId)
 	case Delete:
 		klog.Warningf("deleting tenant %v", e.tenantId)
-		return tc.deleteResources(e.tenantId)
+		return nil
+	}
+	return nil
+}
+
+// onTenantCreate handles the CREATE event of a Tenant.
+func (tc *TenantController) onTenantCreate(tenantId string) error {
+	tenant, err := tc.tenantClient.Tenants().Get(context.TODO(), tenantId, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+
+	if tenant.ObjectMeta.DeletionTimestamp.IsZero() && !util.ContainString(tenant.ObjectMeta.Finalizers, tenantFinalizerKey) {
+		tenant.ObjectMeta.Finalizers = append(tenant.ObjectMeta.Finalizers, tenantFinalizerKey)
+		if _, err := tc.tenantClient.Tenants().Update(context.TODO(), tenant, metav1.UpdateOptions{}); err != nil {
+			return err
+		}
+	}
+
+	return tc.syncResources(tenantId)
+}
+
+// onTenantUpdate handles the UPDATE event of a Tenant.
+func (tc *TenantController) onTenantUpdate(tenantId string) error {
+	tenant, err := tc.tenantClient.Tenants().Get(context.TODO(), tenantId, metav1.GetOptions{})
+	if err != nil {
+		return err
 	}
+
+	if !tenant.ObjectMeta.DeletionTimestamp.IsZero() && util.ContainString(tenant.ObjectMeta.Finalizers, tenantFinalizerKey) {
+		if err := tc.deleteResources(tenantId); err != nil {
+			return err
+		}
+		tenant.ObjectMeta.Finalizers = util.RemoveString(tenant.ObjectMeta.Finalizers, tenantFinalizerKey)
+		if _, err := tc.tenantClient.Tenants().Update(context.TODO(), tenant, metav1.UpdateOptions{}); err != nil {
+			return err
+		}
+	}
+
 	return nil
 }
 
 // deleteResources deletes resources belonging to the tenant from the upstream cluster.
-// TODO delete all cluster-scoped resources belonging to the tenant.
 func (tc *TenantController) deleteResources(tenantId string) error {
-	if tc.upstreamCoreClient == nil || tc.upstreamRbacClient == nil {
-		return errors.New("Can't delete namespaces or RBAC resources with nil client.")
+	if tc.upstreamDiscoveryClient == nil || tc.upstreamDynamicClient == nil {
```

Comment:
Yes, I don't think these clients in proxyConfig will be nil. Because in this case it is already returned when proxyConfig is constructed. I have removed this check in the new commit.